File: app/src/main/java/com/jamieadkins/gwent/data/interactor/CardsInteractor.java
package com.jamieadkins.gwent.data.interactor;

import com.jamieadkins.commonutils.mvp.BaseInteractor;
import com.jamieadkins.commonutils.mvp.BasePresenter;
import com.jamieadkins.gwent.card.CardFilter;
import com.jamieadkins.gwent.data.CardDetails;

import io.reactivex.Completable;
import io.reactivex.Observable;
import io.reactivex.Single;

/**
 * Card manipulation class.
 */

public interface CardsInteractor extends BaseInteractor<BasePresenter> {

    Observable<RxDatabaseEvent<CardDetails>> getCards(CardFilter filter);

    Observable<RxDatabaseEvent<CardDetails>> getCardsIntelligentSearch(CardFilter filter);

    Single<RxDatabaseEvent<CardDetails>> getCard(String id);

    Completable reportMistake(String cardid, String description);

    void removeListeners();

    void setLocale(String locale);
}

File: app/src/main/java/com/jamieadkins/gwent/data/interactor/CardsInteractor.kt
package com.jamieadkins.gwent.data.interactor

import com.jamieadkins.commonutils.mvp.BaseInteractor
import com.jamieadkins.gwent.card.CardFilter
import com.jamieadkins.gwent.data.CardDetails
import com.jamieadkins.gwent.data.CardListResult

import io.reactivex.Completable
import io.reactivex.Flowable
import io.reactivex.Single

/**
 * Card manipulation class.
 */

interface CardsInteractor : BaseInteractor {

    fun getAllCards(): Flowable<Collection<CardDetails>>

    fun getCards(filter: CardFilter): Flowable<Collection<CardDetails>>

    fun getCards(filter: CardFilter?, cardIds: List<String>): Flowable<Collection<CardDetails>>

    fun getCards(filter: CardFilter?, query: String?): Flowable<Collection<CardDetails>>

    fun getCard(id: String): Flowable<CardDetails>
}
